### DISCUSSION TITLE
RTMP-E and Red5 - Need a client to debug

### DISCUSSION THREAD

**Discussion by mondain** [veovera/enhanced-rtmp#34] [General]
If anyone knows of a test app or service I can use to debug my RTMP-E (enhanced-rtmp not RTMPE) implementation in Red5, feel free to share. My search thus far has come up short and I cannot use the rtmfp-cpp testing tools since they don't support regular RTMP as a client. Neither OBS nor builds of ffmpeg appear to send any codec information in the connect call.

**Comment by zenomt** [comment]
it's true that `tcconn` and `tcpublish` in [rtmfp-cpp](https://github.com/zenomt/rtmfp-cpp) only make outbound RTMFP connections. however, `tcrelay` can accept an RTMFP or RTWebSocket connection, and then forward that as RTMP to a server. so i have two multi-hop client setups you could use:
* `tcconn` or `tcpublish` (RTMFP) -> `tcrelay` (RTMP) -> Red5
* [`play-demo.html`](https://github.com/zenomt/rtwebsocket) (RTWebSocket) -> `tcrelay` (RTMP) -> Red5

the latter might even play some video & audio for you. :) i know that AV1 and HEVC work in `play-demo.html` with E-RTMP. while "enhanced audio" (AAC, MP3, Opus, FLAC) should work for playback, i haven't tested it yet, because there's no sample media. Chrome (or a Chromium-based browser like Edge) is required for playback because it uses WebCodecs.

oh yeah, i also don't support E-RTMP "Multitrack", though `tcconn` and `tcrelay` _should_ pass multitrack messages through.

let me know if you need sample command lines to set up `tcrelay` for the above cases.

**Reply by mondain** [reply]
Thanks for the info @zenomt I'll definitely circle back to **rtmfp-cpp** since nothing else sends the `connect` properties I'm trying to verify at the jump of all this effort, which surprises me and makes me assume I've made some mistake.

**Reply by mondain** [reply]
@zenomt I've grabbed **rtwebsocket** and I'm ready for any sample command lines you've got to spare. Everything is local to my dev box and my RTMP server is Red5 of course. Lastly, I looked around your repos and I'd like to give kudos on your errata repo, very nice; I've often wondered why Adobe left so much out and why a fair amount of the info was wrong.

**Reply by zenomt** [reply]
re "so much left out and a fair amount wrong": it was just time constraint. i should _at least_ have been writing a new and independent implementation based solely on the spec while i was editing it, to catch some of the worst errors. it was unfortunately not a priority at the time. i really wish i had completely rewritten it back then instead. the _Errata and Addenda_ is me trying to make amends. :)

these command lines are all to be run from the `rtmfp-cpp/test` directory. i assume Red5 is listening on TCP 1935 on the same computer.

to run `tcrelay` to forward RTMFP to RTMP (to use `tcconn` or `tcpublish`), listening for UDP on `0.0.0.0:1935` and `[::]:1935`, leaving all real-time settings at default, and with enough verboseness to see when connections happen (add more `v`s to get more verboseness):

    $ ./tcrelay -v -46 -i rtmfp -o rtmp localhost

to connect to that with `tcconn` (which will then forward using RTMP to localhost), playing stream `stream` from app `app/instance`:

    $ ./tcconn rtmfp://localhost/app/instance#stream

(add some `-v` before the URI to get some more verbose output from that).

to listen for RTWebSocket on `0.0.0.0:8080` & `[::]:8080` and forward to RTMP on localhost:

    $ ./tcrelay -v -46 -p 8080 -i rtws -o rtmp localhost

to connect to that, open `play-demo.html` in Chrome and **served from a web server** (the way i'm using WebAudio i think requires being served from an HTTP server), change the "Server" URI to be appropriate for your server (something like `ws://localhost:8080/app/instance#stream`), uncheck "tcserver extras" (since you're not connecting to `tcserver`), then hit the connect button and then one of the "Play" buttons. you can open Chrome's JavaScript developer console to see some messages and events as they happen (like the `NetConnection.Connect.Success` result from the connection).

note that if Red5 is fussy about the `tcUrl`, you can override it in the "tcUrl override" box before connecting. this is needed when connecting to Adobe Media Server, which definitely doesn't like the `tcUrl` having a weird scheme like `ws`.

if Chrome isn't running on the same computer as your HTTP server, you'll need to use HTTPS probably. if that's too much hassle, email me and i'll give you an HTTP URL for `play-demo.html` running on my home server. and if Chrome isn't running on the same computer as `tcrelay` (so it's not `localhost`), you'll probably need to run that behind a TLS reverse proxy. i recommend if at all possible to run Chrome and `tcrelay` on the same computer to avoid that specific hassle right now. :)

note that `tcrelay`, `tcconn`, `tcpublish`, and `tcserver` will all give you their help if run without arguments.

**Reply by zenomt** [reply]
to make it easier for `play-demo.html`, i just published that repo as GitHub Pages. so you should be able to open https://zenomt.github.io/rtwebsocket/play-demo.html and not bother hosting it on your own web server (though you can of course).

**Reply by mondain** [reply]
Thanks, for all your help, I sorta got it working, I have at the very least a correct **connect** from an enhanced RTMP client now. I'll have to come back to this later; here's what your implementation provided, which is what I expected:

```
2024-06-22 15:10:15,290 [NioProcessor-3] DEBUG org.red5.server.net.rtmp.codec.RTMPProtocolDecoder - Num connection params: 9
2024-06-22 15:10:15,290 [NioProcessor-3] DEBUG org.red5.server.net.rtmp.codec.RTMPProtocolDecoder -  > objectEncoding: 0.0
2024-06-22 15:10:15,290 [NioProcessor-3] DEBUG org.red5.server.net.rtmp.codec.RTMPProtocolDecoder -  > tcUrl: rtmp://10.0.0.35:1935/live
2024-06-22 15:10:15,290 [NioProcessor-3] DEBUG org.red5.server.net.rtmp.codec.RTMPProtocolDecoder -  > app: live
2024-06-22 15:10:15,290 [NioProcessor-3] DEBUG org.red5.server.net.rtmp.codec.RTMPProtocolDecoder -  > videoCodecs: 128.0
2024-06-22 15:10:15,290 [NioProcessor-3] DEBUG org.red5.server.net.rtmp.codec.RTMPProtocolDecoder -  > audioCodecs: 65535.0
2024-06-22 15:10:15,290 [NioProcessor-3] DEBUG org.red5.server.net.rtmp.codec.RTMPProtocolDecoder -  > fourCcList: [av01, vp09, hvc1, avc1, Opus, .mp3, fLaC, mp4a]
2024-06-22 15:10:15,290 [NioProcessor-3] DEBUG org.red5.server.net.rtmp.codec.RTMPProtocolDecoder -  > capsEx: 0.0
2024-06-22 15:10:15,290 [NioProcessor-3] DEBUG org.red5.server.net.rtmp.codec.RTMPProtocolDecoder -  > audioFourCcInfoMap: {Opus=1.0, .mp3=1.0, fLaC=1.0, mp4a=1.0}
2024-06-22 15:10:15,290 [NioProcessor-3] DEBUG org.red5.server.net.rtmp.codec.RTMPProtocolDecoder -  > videoFourCcInfoMap: {av01=1.0, vp09=1.0, hvc1=1.0, avc1=1.0}
2024-06-22 15:10:15,290 [NioProcessor-3] DEBUG org.red5.server.net.rtmp.codec.RTMPProtocolDecoder - Num params: 0
2024-06-22 15:10:15,291 [RTMPConnectionExecutor-1] DEBUG org.red5.server.net.rtmp.BaseRTMPHandler - getHostname - url: rtmp://10.0.0.35:1935/live
```
One of these days I'll kill all that reading numbers as doubles, but for now its usable.

**Reply by zenomt** [reply]
> One of these days I'll kill all that reading numbers as doubles, but for now its usable.

to be fair, AMF Numbers are always floating point. that's a "printing" problem. if it was C/C++ i'd say "use %g in your printf". i don't know about Java. :)

note that `tcconn` will send the wildcard `*` for `fourCcList`, `audioFourCcInfoMap`, and `videoFourCcInfoMap` with value 5 (saying that it can forward and decode all codecs), plus send 65535 for `videoCodecs` and `audioCodecs` (saying that it can accept any traditional codec). it should be obvious from the source for both `play-demo.html` and `tcconn.cpp` where to change those, if you want to experiment with how Red5 behaves if you do or don't send a particular codec, bit flag, wildcard, etc.

**Reply by zenomt** [reply]
2024-06-22 15:10:15,290 [NioProcessor-3] DEBUG org.red5.server.net.rtmp.codec.RTMPProtocolDecoder - Num params: 0

i assume "`Num params: 0`" is referring to the number of `connect` arguments after the Command Object. FYI with `tcconn`, `tcpublish`, and `play-demo.html` you can add extra String arguments to `connect` by putting them in the `userinfo` of the URI:

    rtmfp://arg1:arg2:arg3@localhost/live#streamName
    ws://arg1:arg2::arg4@localhost:8080/live#streamName

more info about the `userinfo` and `fragment` productions at https://zenomt.github.io/rtmp-errata-addenda/rtmp-uris

**Reply by mondain** [reply]
Yes, those are the extra args/params one can pass via the uri.

**Reply by veovera** [reply]
> Thanks, for all your help, I sorta got it working,

Great work to all involved. @mondain can you describe your setup in more detail so perhaps others can leverage your work? thx!

**Reply by mondain** [reply]
@veovera I'll be sure to post once my implementation in Red5 is ready to go; I commend you on your easy-to-follow specification. I would however like to see more low level detail on the multichannel section, specifically, but not limited to codec per track determination.

**Comment by zenomt** [comment]
note that `tcrelay` can turn _any_ RTMP server into an RTMFP (or RTWebSocket) server. :)

**Comment by zenomt** [comment]
oh, i noticed i hadn't updated `play-demo.html` to send the latest fourCcs in `fourCcList`, nor to send `audioFourCcInfoMap` and `videoFourCcInfoMap`. it does now.

**Comment by mondain** [comment]
@veovera which version of FFMPEG supports the current spec?

**Reply by mondain** [reply]
I have found that OBS beta 30.2.0 rc1 supports v2, but I don't yet know whether or not it will multi-track.

**Reply by mondain** [reply]
To answer my own question; OBS beta does not do multitrack nor Opus yet and is feature frozen.

**Comment by mondain** [comment]
@veovera @zenomt Quick question, when I have incoming enhanced audio or video that is not multitrack, will I always get the four byte fourcc code preceding every packet type's data contents? SequenceStart, SequenceEnd, CodedFrames, CodedFramesX, and Metadata; the last is the one I haven't seen yet, so it seems quicker to ask.

**Reply by zenomt** [reply]
that's my reading of the spec. my implementation treats a message that's marked as "enhanced", but that's not at least 5 bytes long, as "not enhanced": https://github.com/zenomt/rtmfp-cpp/blob/63ee4f3f98a6e1372af7e25453d388463bf592fd/src/TCMessage.cpp#L143-L146

**Reply by veovera** [reply]
> when I have incoming enhanced audio or video that is not multitrack, will I always get the four byte fourcc code preceding every packet type's data contents?

In most cases that is correct. Please see the pseudocode 

- [for audio](https://github.com/veovera/enhanced-rtmp/blob/17379dd712f8d335b6d90e1f0046b7d21ed6961b/docs/enhanced/enhanced-rtmp-v2.md?plain=1#L652)
- [for video](https://github.com/veovera/enhanced-rtmp/blob/17379dd712f8d335b6d90e1f0046b7d21ed6961b/docs/enhanced/enhanced-rtmp-v2.md?plain=1#L990)

As you can see when in enhanced mode and fetching video metadata you don't fetch a FOURCC value.

**Reply by zenomt** [reply]
@veovera two things:

first: i didn't notice until now that the treatment for "enhanced video with frametype 'command'" is different from E-RTMP v1. i wish i had noticed this before; i would have had discussion items. i assume this change is so that smart-seeking a multi-track stream (which might have different codecs for different tracks) can only have all tracks start/end smart-seek together.

second: my reading is that _only_ an enhanced video message that's both **not** metadata and is also a video command will not have a FourCC. a message that **is** metadata (regardless of if it's a video command), or a message that isn't metadata but also isn't a command, will have a FourCC.

**Reply by veovera** [reply]
E-RTMP v2 is currently in alpha, allowing for potential breaking changes. This flexibility enables us to refine and evolve the protocol during this phase. However, as we move towards beta, such changes will be stabilized. Note: so far there have not been any breaking changes in E-RTMP v2 alpha.

Regarding the treatment of "enhanced video with frametype 'command,'" it's worth noting that in both E-RTMP v1 and E-RTMP v2, our approach remains consistent. When `videoFrameType` is `VideoFrameType.Command` (i.e., 5), the payload does not contain video data. Instead, the VideoTagHeader is followed by a UI8 indicating:

- 0 for the start of a client-side seeking video frame sequence
- 1 for the end of such a sequence

This approach ensures uniformity across versions. While our pseudocode may not always cover every detail exhaustively, it aims to comprehensively address scenarios involving `VideoFrameType.Command` in E-RTMP v2.

In legacy RTMP, if `FrameType` is command, instead of the video payload the message stream includes a UI8 indicating the command type, similar to the approach described above.

**Comment by mondain** [comment]
@veovera the more I implement to the spec, the more I like it; thanks for your efforts to extend the life of RTMP.

**Reply by veovera** [reply]
Thank you for the feedback, this makes my day! I am looking forward to seeing your resulting service. Keep us posted...

**Comment by mondain** [comment]
@veovera I grabbed FFMPEG latest and built with all the appropriate codecs, but I cannot get VP8 to work; the v2 spec says the fourcc is generated with `vp08` and I've tried that and `vp80` with neither working: ```Video: none (vp80 / 0x30387076), none``` I also checked the VLC header source for clues and they're using uppercase ASCII as `VP80` who's the most correct here? https://github.com/videolan/vlc/blob/master/src/misc/fourcc_list.h
Lastly, seems I'll run into the same issue with VP9 and AV1..

**Reply by mondain** [reply]
It would seem that ffmpeg doesn't support vp8 in flv; `[flv @ 0x55ce4143dbc0] Video codec vp8 not compatible with flv`

**Reply by veovera** [reply]
> who's the most correct here?

As far as FOURCC values we use the values defined in the specs below

- https://www.webmproject.org/vp9/mp4/ (for "vp08" and "vp09")
- https://aomediacodec.github.io/av1-isobmff/v1.3.0.html#av1codecconfigurationbox-definition (for "av01")

FFmpeg uses the same values for e-rtmp/flv see <[link](https://github.com/FFmpeg/FFmpeg/blob/b6c43328eedfdc755e8ee378f8af2034ac67f62d/libavformat/flvenc.c#L527)>

**Reply by veovera** [reply]
> It would seem that ffmpeg doesn't support vp8 in flv; `[flv @ 0x55ce4143dbc0] Video codec vp8 not compatible with flv`

It looks like FFmpeg chose not to support vp8. Today I don't believe vp8 is a popular codec. We chose to add vp8 to the spec because webRTC officially supports this codec.

**Reply by mondain** [reply]
It may not be all that popular, but a lot of Android devices still have it and its a requirement in the WebRTC spec. I'll work around it and maybe do a PR for ffmpeg later; thanks for the updates.

**Reply by veovera** [reply]
> its a requirement in the WebRTC spec

Agree 100%, that's why we included it! Post your PR link here if you get a chance. thx!

**Comment by mondain** [comment]
@veovera here are my current example FLV files for AV1 and VP8; this repo also includes the FFmpeg patch that I'm waiting to submit, once they let me on the list. https://github.com/mondain/mediafiles

**Reply by zenomt** [reply]
i'm able to play your AV1 sample; however, i need a little bit of playback buffer or the framerate looks very low (i suspect from out-of-order frames or something). that is, playback with almost 0 buffer looks like about 5 frames per second, but if i play back with a half-second buffer, the video looks smooth.

i need to go add support for VP8 at least on my playback side. it looks like Chrome has WebCodecs support for that, so it should just be a few lines of code. i should be able to confirm playback pretty soon.

**Reply by zenomt** [reply]
i can play the VP8 sample now (changes not pushed to github yet). however, it looks like audio and video are a little bit out-of-sync in the VP8 sample, with audio maybe 100-200 ms behind (just based on my eye/ear, not actually measuring).

**Reply by zenomt** [reply]
VP8 support [(+3 lines of code)](https://github.com/zenomt/rtwebsocket/commit/14066c68ae243499b45ae96f717efd5f8de418a8) added to https://github.com/zenomt/rtwebsocket . i also added the codec ID to https://github.com/zenomt/rtmfp-cpp but that's not necessary for tcpublish/tcconn/tcserver to handle it.

**Reply by zenomt** [reply]
actually, it looks like the AV1 sample might also have the same A/V sync issue; however, that sample is much shorter and there's only one good part (the neck crack) to judge sync (although the "ahhh" after smelling the flowers is also off, but that one it's harder to tell).

**Reply by mondain** [reply]
These are just base recordings until I get all the other stuff sorted; I'm not using the comp offset timing data at the moment for those codecs that supply it.

**Comment by koolscooby** [comment]
A few resources for folks :
* ertmp-toolbox: Simple rust utility that can read and push eFLV files https://github.com/kc5nra/ertmp-toolbox 
* FFmpeg: Branch that has Enhanced RTMP Multitrack FLV support https://github.com/BtbN/FFmpeg/tree/enhanced-flv
* Wireshark: v4.4+ support dissecting/parsing Enhanced RTMP Multitrack connections
* OBS Studio: v30.2+ support Twitch Enhanced Broadcasting powered by Enhanced RTMP Multitrack Feature. A CLI option `--enable-multitrack-video-dev` lets you experiment with custom OBS configurations for testing streaming to services other than Twitch and provides an optional checkbox to record an eFLV file.

**Reply by veovera** [reply]
@koolscooby Thank you for sharing!
